recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py

```diff
@@ -0,0 +1,157 @@
+
+def get_kernel_version():
+    return run_cmd(['uname', '-r'])
+
+def get_architecture():
+    return run_cmd(['uname', '-m'])
+
+def get_os_version():
+    with open('/etc/os-release', 'rb') as file, mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) as mfile:
+        build_id = re.search(br'BUILD_ID=\"(.*)\"', mfile)
+        return build_id.group(1).decode()
+
+def is_next_kernel():
+    kernel_version = get_kernel_version()
+    return kernel_version.endswith('-next')
+
+def get_dmesg_log():
+    dmesg_log = run_cmd(['dmesg'])
+    return dmesg_log
+
+def upload_log(db, dmesg_log):
+    data = {}
+    data['kernel_version'] = get_kernel_version()
+    data['kernel_type'] = 'next' if is_next_kernel() else 'current'
+    data['device_desc'] = get_device_desc()
+    data['architecture'] = get_architecture()
+    data['os_version'] = get_os_version()
+    data['date'] = str(datetime.datetime.now())
+
+    header = ''
+    for key, val in data.items():
+        header = header + '# {}: {}\n'.format(key, val)
+
+    data['dmesg_log'] = header + dmesg_log
+
+    record = db.insert(data)
+    print('INFO: Uploaded current dmesg log from {} with _id {}'.format(data['date'], record.inserted_id))
+
+def strip_headers(dmesg_log):
+    output = ''
+    for line in dmesg_log.splitlines():
+        if not line.startswith('#'):
+            output = output + line + '\n'
+    return output
+
+def get_old_dmesg_log(db):
+    query = {}
+    query['kernel_version'] = {'$ne': get_kernel_version()}
+    query['kernel_type'] = 'next' if is_next_kernel() else 'current'
+    query['device_desc'] = get_device_desc()
+    results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
+    if not results:
+        return None
+
+    for result in results:
+        print('INFO: Using previous dmesg log from {} with _id {}'.format(result['date'], result['_id']))
+        dmesg_log = result['dmesg_log']
+        return strip_headers(dmesg_log)
+
+    print('INFO: No suitable previous dmesg log found')
+    return ''
+
+def strip_timestamps(log):
+    return re.sub(r'^\[.+?\] ', '', log, flags=re.MULTILINE)
+
+replacement_patterns = [
+        [r'Linux version \S+', 'Linux version '],
+        [r'root=PARTUUID=.+? ', 'root=PARTUUID= '],
+        [r'setting system clock to .*', 'setting system clock to '],
+        [r'sched_clock: Marking stable.*', 'sched_clock: Marking stable '],
+        [r'udevd\[\d+\]: ', 'udevd[]: '],
+        [r'audit\(\d+\.\d+:\d+\):', 'audit():']
+]
+
+def strip_known_differences(log):
+    for pattern in replacement_patterns:
+        log = re.sub(pattern[0], pattern[1], log, flags=re.MULTILINE)
+    return log
+
+def prepare_log_for_diff(log):
+    log = strip_timestamps(log)
+    log = strip_known_differences(log)
+    log = log.splitlines()
+    log.sort()
+    return log
+
+def diff_logs(current_log, old_log):
+    current_log = prepare_log_for_diff(current_log)
+    old_log = prepare_log_for_diff(old_log)
```

Comment:
Is the test going to fail if we don't have a 'old_log' to compare against? Maybe that's OK for the first test run.

Reply:
Yes, if there isn't an old log, `old_log` is empty string and the test will fail as diff will be non-empty.

I'm planning to seed the db with 22.5 BSI's kernel log. But the test will likely fail - at least the first time anyway as there are so many changes and it is not possible to put all those in allow list.